File: Frontend/SeatifyFrontend/src/app/services/checkin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { ConfigService } from './config.service';

export interface CheckInValidateRequest {
  payload: string;
}

export interface CheckInConfirmRequest {
  ticketId: string;
}

export enum TicketStatus {
  Valid = 0,
  AlreadyUsed = 1,
  Invalid = 2
}

export interface ReservationInfo {
  customerName: string;
  customerEmail: string;
}

export interface SeatInfo {
  section: string;
  row: number;
  number: number;
  seatLabel?: string;
}

export interface EventInfo {
  title: string;
  startTime: string;
}

export interface CheckInResult {
  ticketId: string;
  status: TicketStatus;
  statusMessage: string;
  reservation?: ReservationInfo;
  seat?: SeatInfo;
  event?: EventInfo;
}

@Injectable({
  providedIn: 'root'
})
export class CheckInService {

  private apiUrl = `${environment.baseApiUrl}/api/checkin`;

  private readonly checkInPath = '/api/checkin';

  constructor(
    private http: HttpClient,
    private configService: ConfigService
  ) { }

  private api(path: string): string {
    return `${this.configService.cfg.baseApiUrl}${path}`;
  }

  validateTicket(payload: string): Observable<CheckInResult> {
    return this.http.post<CheckInResult>(`${this.api(this.checkInPath)}/validate`, { payload });
  }

  confirmCheckIn(ticketId: string): Observable<CheckInResult> {
    return this.http.post<CheckInResult>(`${this.api(this.checkInPath)}/confirm`, { ticketId });
  }
}

File: Backend/SeatifyBackend/Logic/Services/CheckInService.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Data;
using Entities.Dtos.CheckIn;
using Logic.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Logic.Services
{
    public class CheckInService : ICheckInService
    {
        private readonly AppDbContext _context;

        public CheckInService(AppDbContext context)
        {
            _context = context;
        }

        public async Task<CheckInResult> ValidateTicketAsync(string payload)
        {
            var ticketId = payload; // Assuming payload is the ReservationSeat.Id for now
            var result = new CheckInResult { TicketId = ticketId };

            var reservationSeat = await _context.ReservationSeats
                .Include(rs => rs.Reservation)
                .ThenInclude(r => r.EventOccurrence)
                .ThenInclude(e => e.Event)
                .Include(rs => rs.Seat)
                .ThenInclude(s => s.Sector)
                .FirstOrDefaultAsync(rs => rs.Id == ticketId);

            if (reservationSeat == null)
            {
                result.Status = TicketStatus.Invalid;
                result.StatusMessage = "Ticket not found.";
                return result;
            }

            result.Reservation = new ReservationInfo
            {
                CustomerName = reservationSeat.Reservation.CustomerName,
                CustomerEmail = reservationSeat.Reservation.CustomerEmail
            };

            result.Seat = new SeatInfo
            {
                Section = reservationSeat.Seat.Sector?.Name ?? "General",
                Row = reservationSeat.Seat.Row,
                Number = reservationSeat.Seat.Column
            };

            var occurrence = reservationSeat.Reservation.EventOccurrence;
            if (occurrence != null)
            {
                result.Event = new EventInfo
                {
                    Title = occurrence.Event?.Name ?? "Unknown Event",
                    StartTime = occurrence.StartsAtUtc
                };
            }

            if (reservationSeat.IsCheckedIn)
            {
                result.Status = TicketStatus.AlreadyUsed;
                result.StatusMessage = $"Ticket was already checked in at {reservationSeat.CheckInTimeUtc}.";
            }
            else
            {
                result.Status = TicketStatus.Valid;
                result.StatusMessage = "Ticket is valid for check-in.";
            }

            return result;
        }

        public async Task<CheckInResult> ConfirmCheckInAsync(string ticketId)
        {
            var result = await ValidateTicketAsync(ticketId);
            
            if (result.Status == TicketStatus.Valid)
            {
                var reservationSeat = await _context.ReservationSeats.FindAsync(ticketId);
                if (reservationSeat != null)
                {
                    reservationSeat.IsCheckedIn = true;
                    reservationSeat.CheckInTimeUtc = DateTime.UtcNow;
                    await _context.SaveChangesAsync();
                    
                    result.StatusMessage = "Check-in successful.";
                }
            }

            return result;
        }
    }
}
